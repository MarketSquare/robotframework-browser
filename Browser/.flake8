[flake8]
exclude =
   __pycache__,
   generated,
	 wrapper
ignore = E501, W503, B008
