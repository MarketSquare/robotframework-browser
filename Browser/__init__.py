# Copyright 2020-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .browser import Browser
from .utils.data_types import (
    AssertionOperator,
    ColorScheme,
    ElementState,
    KeyboardModifier,
    MouseButton,
    RequestMethod,
    SelectAttribute,
    SupportedBrowsers,
    ViewportDimensions,
)
from .version import __version__ as VERSION

__version__ = VERSION
__all__ = [
    "AssertionOperator",
    "ElementState",
    "ColorScheme",
    "ViewportDimensions",
    "SupportedBrowsers",
    "SelectAttribute",
    "KeyboardModifier",
    "MouseButton",
    "RequestMethod",
    "Browser",
]
