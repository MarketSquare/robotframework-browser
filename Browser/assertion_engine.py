# Copyright 2020-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import time
from typing import Optional, Union

import wrapt  # type: ignore
from assertionengine import AssertionOperator
from typing_extensions import get_args, get_origin

from .utils import logger


def assertion_operator_is_set(wrapped, args, kwargs):
    assertion_operator = None
    assertion_op_name = None
    assertion_op_index = None
    for index, (_arg, typ) in enumerate(wrapped.__annotations__.items()):
        if get_origin(typ) is Union and AssertionOperator in get_args(typ):
            assertion_op_index = index
            break
        if typ is AssertionOperator:
            assertion_op_index = index
            break
    if assertion_op_index is not None:
        if len(args) > assertion_op_index:
            assertion_operator = args[assertion_op_index]
        elif assertion_op_name in kwargs:
            assertion_operator = kwargs[assertion_op_name]
    return assertion_operator


@wrapt.decorator
def with_assertion_polling(wrapped, instance, args, kwargs):
    start = time.time()
    timeout = instance.timeout / 1000
    retry_assertions_until = instance.retry_assertions_for / 1000
    retries_start: Optional[float] = None
    tries = 1
    try:
        logger.stash_this_thread()
        while True:
            try:
                return wrapped(*args, **kwargs)
            except AssertionError as e:
                if retries_start is None:
                    retries_start = time.time()
                elapsed = time.time() - start
                elapsed_retries = time.time() - retries_start
                if elapsed >= timeout or elapsed_retries >= retry_assertions_until:
                    raise e
                tries += 1
                if timeout - elapsed > 0.01:  # noqa: PLR2004
                    time.sleep(0.01)
                logger.clear_thread_stash()
    finally:
        logger.flush_and_delete_thread_stash()
        if retry_assertions_until and assertion_operator_is_set(wrapped, args, kwargs):
            now = time.time()
            logger.debug(
                f"Assertion polling statistics:\n"
                f"First element asserted in: {(retries_start or now) - start} seconds\n"
                f"Total tries: {tries}\n"
                f"Elapsed time in retries {now - (retries_start or now)} seconds"
            )
