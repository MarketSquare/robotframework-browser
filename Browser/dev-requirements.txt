invoke >= 2.1.1
pytest >= 7.3.1
pytest-watch >= 4.2.0
black >= 23.7.0
mypy >= 1.4.1
mypy-protobuf >= 3.4.0
ruff>=0.0.277
wheel>=0.40.0
robotframework-pabot >= 2.16.0
twine >= 4.0.2
robotstatuschecker >= 3.0.1
rellu >= 0.7
approvaltests == 8.3.1
pytest-mock==3.11.1
Pillow==10.0.0
# Required for injecting google analytics tags on release
beautifulsoup4 >= 4.12.2
psutil >= 5.9.4
coverage >= 7.2.3
robotframework-tidy >= 4.4.0
python-dateutil >= 2.8.2
robotframework-crypto >= 0.3.0
# Include normal dependencies from requirements.txt. Makes it possible to use
# requirements-dev.txt as a single requirement file in PyCharm and other IDEs.
-r requirements.txt
