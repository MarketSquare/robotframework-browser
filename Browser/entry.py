# Copyright 2020-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import platform
import subprocess
import sys
from pathlib import Path
from subprocess import DEVNULL, PIPE, STDOUT, CalledProcessError, Popen

USAGE = """USAGE
  rfbrowser [command]
AVAILABLE COMMANDS
  init  Install required nodejs dependencies
    OPTIONS:
        --skip-browsers

"""


def run():
    if len(sys.argv) < 2:
        print(USAGE)
        sys.exit(1)
    cmd = sys.argv[1]
    if cmd == "init":
        arg2 = sys.argv[2] if len(sys.argv) >= 3 else ""
        rfbrowser_init(arg2 == "--skip-browsers")
    else:
        print(f"Invalid command `{cmd}`")
        print(USAGE)
        sys.exit(64)


def rfbrowser_init(skip_browser_install: bool):
    print("Installing node dependencies...")
    installation_dir = Path(__file__).parent / "wrapper"

    if not (installation_dir / "package.json").is_file():
        print(
            f"Installation directory `{installation_dir}` does not contain the required package.json "
            + "\nPrinting contents:"
        )
        for root, _dirs, files in os.walk(installation_dir):
            level = root.replace(installation_dir.__str__(), "").count(os.sep)
            indent = " " * 4 * (level)
            print("{}{}/".format(indent, os.path.basename(root)))
            subindent = " " * 4 * (level + 1)
            for f in files:
                print("{}{}".format(subindent, f))
        raise RuntimeError("Could not find robotframework-browser's package.json")
    if not os.access(installation_dir, os.W_OK):
        sys.tracebacklimit = 0
        raise RuntimeError(
            f"`rfbrowser init` needs write permissions to {installation_dir}"
        )

    print("Installing rfbrowser node dependencies at {}".format(installation_dir))

    try:
        # This is required because weirdly windows doesn't have `npm` in PATH without shell=True.
        # But shell=True breaks our linux CI
        shell = True if platform.platform().startswith("Windows") else False
        subprocess.run(["npm", "-v"], stdout=DEVNULL, check=True, shell=shell)
    except (CalledProcessError, FileNotFoundError, PermissionError) as exception:
        print(
            "Couldn't execute npm. Please ensure you have node.js and npm installed and in PATH."
            "See https://nodejs.org/ for documentation"
        )
        sys.exit(exception)

    if skip_browser_install:
        os.environ["PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD"] = "1"
    else:
        os.environ["PLAYWRIGHT_BROWSERS_PATH"] = "0"

    process = Popen(
        "npm install --production",
        shell=True,
        cwd=installation_dir,
        stdout=PIPE,
        stderr=STDOUT,
    )

    while process.poll() is None:
        if process.stdout:
            output = process.stdout.readline()
            print(output.decode("utf-8"))

    if process.returncode != 0:
        raise RuntimeError(
            "Problem installing node dependencies."
            + f"Node process returned with exit status {process.returncode}"
        )

    print("rfbrowser init completed")


if __name__ == "__main__":
    run()
