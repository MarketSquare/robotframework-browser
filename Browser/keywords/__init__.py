# Copyright 2020-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .browser_control import Control
from .cookie import Cookie
from .device_descriptors import Devices
from .evaluation import Evaluation
from .getters import Getters
from .interaction import Interaction
from .network import Network
from .playwright_state import PlaywrightState
from .promises import Promises
from .runonfailure import RunOnFailureKeywords
from .strict_mode import StrictMode
from .waiter import Waiter
from .webapp_state import WebAppState

__all__ = [
    "Control",
    "Cookie",
    "Devices",
    "Getters",
    "Evaluation",
    "Interaction",
    "Network",
    "PlaywrightState",
    "Promises",
    "RunOnFailureKeywords",
    "StrictMode",
    "Waiter",
    "WebAppState",
]
