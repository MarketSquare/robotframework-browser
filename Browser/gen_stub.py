# Copyright 2020-     Robot Framework Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from datetime import timedelta
from typing import Any

from robotlibcore import KeywordBuilder  # type: ignore

import Browser


def is_named_method(keyword_name: str) -> bool:
    keyword_attribute = br.attributes[keyword_name]
    return (
        keyword_attribute.robot_name is not None
        and keyword_attribute.robot_name == keyword_name
    )


def get_method_name_for_keyword(keyword_name: str) -> str:
    if is_named_method(keyword_name):
        for key in br.attributes.keys():
            if key != keyword_name and keyword_name == br.attributes[key].robot_name:
                return key
    return keyword_name


def get_type_string_from_type(argument_type: type) -> str:
    if hasattr(argument_type, "__name__"):
        return argument_type.__name__
    else:
        arg_type_str = str(argument_type.__repr__()).lstrip("typing.")
        return arg_type_str.replace("NoneType", "None")


def get_type_string_from_argument(argument_string: str, argument_types: dict) -> str:
    agrument_name = argument_string.lstrip("*")
    if agrument_name in argument_types:
        return get_type_string_from_type(argument_types[agrument_name])
    return ""


def get_function_list_from_keywords(keywords):
    functions = list()
    for keyword in keywords:
        method_name = get_method_name_for_keyword(keyword)
        keyword_arguments = br.get_keyword_arguments(keyword)
        keyword_types = br.get_keyword_types(keyword)
        functions.append(keyword_line(keyword_arguments, keyword_types, method_name))
    functions.sort()
    return functions


def keyword_line(keyword_arguments, keyword_types, method_name) -> str:
    arguments_list = list()
    for argument in keyword_arguments:
        if isinstance(argument, tuple):
            arg_str = argument[0]
            default_value = argument[1]
            arg_type_str = get_type_string_from_argument(arg_str, keyword_types)
            if arg_type_str:
                if default_value is None:
                    arg_type_str = f"Optional[{arg_type_str}]"
                arg_str = f"{arg_str}: {arg_type_str}"
            if isinstance(default_value, str):
                default_value = f"'{default_value}'"
            if isinstance(default_value, timedelta):
                default_value = f"timedelta(seconds={default_value.total_seconds()})"
            arg_str = f"{arg_str} = {default_value}"
        else:
            arg_str = argument
            arg_type_str = get_type_string_from_argument(arg_str, keyword_types)
            if arg_type_str:
                arg_str = arg_str + f": {arg_type_str}"
        arguments_list.append(arg_str)
    arguments_string = (
        f", {', '.join(arguments_list)}" if len(arguments_list) > 0 else ""
    )
    return f"    def {method_name}(self{arguments_string}): ...\n"


br: Any = Browser.Browser()
function_list = get_function_list_from_keywords(br.get_keyword_names())


pyi_boilerplate = """\
import datetime
from concurrent.futures import Future
from datetime import timedelta
from typing import (
    Any,
    Dict,
    List,
    Optional,
    Union,
    Tuple,
)

from .utils.data_types import *


class Browser:
    timeout: Any = ...
"""

pyi_non_kw_methods = """\
    def get_timeout(self, timeout: Union[timedelta, None]) -> float: ...
    def convert_timeout(self, timeout: Union[timedelta, float]) -> float: ...
    def millisecs_to_timestr(self, timeout: float)  -> str: ...

"""

init_method = KeywordBuilder.build(br.__init__)
with open("Browser/__init__.pyi", "w") as stub_file:
    stub_file.write(pyi_boilerplate)
    init_string = keyword_line(
        init_method.argument_specification, init_method.argument_types, "__init__"
    )
    # init methods argument types that are contained within higher order types don't get
    # import syntax stripped correctly so we fix them manually here
    init_string = init_string.replace(
        "Dict[Browser.utils.data_types.SupportedBrowsers, str]",
        "Dict[SupportedBrowsers, str]",
    )

    stub_file.write(init_string)
    stub_file.writelines(function_list)
    stub_file.write(pyi_non_kw_methods)
