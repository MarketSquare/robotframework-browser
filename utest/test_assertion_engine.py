import pytest
from approvaltests import verify_all  # type: ignore

from Browser.assertion_engine import (
    verify_assertion,
    AssertionOperator,
    with_assertion_polling,
)
from robot.libraries.BuiltIn import EXECUTION_CONTEXTS  # type: ignore


def test_equals():
    results = [
        _validate_operator(AssertionOperator["=="], "actual", "actual"),
        _validate_operator(AssertionOperator["=="], "actual", "unexpected"),
        _validate_operator(AssertionOperator["=="], 1, "1"),

    ]
    verify_all("Test equals", results)


def test_not_equals():
    results = [
        _validate_operator(AssertionOperator["!="], "actual", "expected"),
        _validate_operator(AssertionOperator["!="], "actual", "actual"),
    ]
    verify_all("Not equal", results)


def test_contains():
    results = [
        _validate_operator(AssertionOperator["contains"], "actual", "ctua"),
        _validate_operator(AssertionOperator["contains"], "actual", "nope"),
        _validate_operator(AssertionOperator["*="], "actual", "tual"),
        _validate_operator(AssertionOperator["*="], "actual", "nope"),
    ]
    verify_all("Contains", results)


class FakeBrowser:
    timeout = 100
    counter = 1
    retry_assertions_for = 30

    @with_assertion_polling
    def is_three(self, value):
        return verify_assertion(value, AssertionOperator["=="], 3)

    @with_assertion_polling
    def second_run_success(self):
        current = self.counter
        self.counter += 1
        return verify_assertion(current, AssertionOperator["=="], 2)


def test_with_assertions_polling():
    fb = FakeBrowser()
    results = [
        fb.is_three(3),
        _method_validator(fb.is_three, 2),
        fb.second_run_success(),
    ]
    verify_all("Assertion polling", results)


def test_without_assertions_polling():
    fb = FakeBrowser()
    fb.retry_assertions_for = 0
    results = [
        fb.is_three(3),
        _method_validator(fb.is_three, 2),
        _method_validator(fb.second_run_success)
    ]
    verify_all("No polling", results)


def test_greater():
    results = [
        _validate_operator(AssertionOperator["<"], 1, 2),
        _validate_operator(AssertionOperator["<"], 1, 0),
    ]
    verify_all("Greater", results)


def test_less():
    results = [
        _validate_operator(AssertionOperator[">"], 2, 1),
        _validate_operator(AssertionOperator[">"], 2, 3),
    ]
    verify_all("Less", results)


def test_greater_or_equal():
    results = [
        _validate_operator(AssertionOperator["<="], 1, 2),
        _validate_operator(AssertionOperator["<="], 1, 0),
        _validate_operator(AssertionOperator["<="], 1, 1),
    ]
    verify_all("greater or equal", results)


def test_less_or_equal():
    results = [
        _validate_operator(AssertionOperator[">="], 2, 2),
        _validate_operator(AssertionOperator[">="], 2, 3),
        _validate_operator(AssertionOperator[">="], 2, 1),
    ]
    verify_all("less or equal", results)


def test_match():
    results = [
        _validate_operator(AssertionOperator["matches"], "Actual", "(?i)actual"),
        _validate_operator(AssertionOperator["matches"], "Actual", "/(\\d)+/"),
        _validate_operator(AssertionOperator["matches"], "Actual", "^Act"),
        _validate_operator(AssertionOperator["matches"], "Actual", "/(\\d)+/"),
        _validate_operator(AssertionOperator["matches"], "Actual", "ual$"),
        _validate_operator(AssertionOperator["matches"], "Actual\nmultiline", "(?m)Actual\nmultiline$"),
        _validate_operator(AssertionOperator["matches"], "Actual\nmultiline", "/(\\d)+/"),

    ]
    verify_all("match", results)


@pytest.fixture()
def with_suite():
    def ns():
        pass

    ns.variables = lambda: 0
    ns.variables.current = lambda: 0
    ns.variables.current.store = lambda: 0
    EXECUTION_CONTEXTS.start_suite("suite", ns, lambda: 0)
    yield
    EXECUTION_CONTEXTS.end_suite()


def test_validate(with_suite):
    results = [
        _validate_operator(AssertionOperator("validate"), 1, "0 < value < 2"),
        _validate_operator(AssertionOperator("validate"), 1, "value == 'hello'"),
    ]
    verify_all("validate", results)


def test_then(with_suite):
    then_op = AssertionOperator["then"]
    results = [
        verify_assertion(8, then_op, "value + 3") == 11,
        verify_assertion(2, then_op, "value + 3") == 5,
        verify_assertion("René", then_op, "'Hello ' + value + '!'") == "Hello René!"
    ]
    verify_all("then", results)


def test_start_with():
    results = [
        _validate_operator(AssertionOperator["^="], "Hello Robots", "Hello"),
        _validate_operator(AssertionOperator["^="], "Hello Robots", "Robots"),
        _validate_operator(AssertionOperator["should start with"], "Hello Robots", "Hello"),
        _validate_operator(AssertionOperator["should start with"], "Hello Robots", "Robots"),
        _validate_operator(AssertionOperator["^="], "Hel[4,5]?[1-9]+ Robots", "Hel[4,5]?[1-"),
        _validate_operator(AssertionOperator["^="], "Hel[4,5]?[1-9]+ Robots", ".*"),

    ]
    verify_all("start with", results)


def test_ends_with():
    results = [
        _validate_operator(AssertionOperator["$="], "Hello Robots", "Robots"),
        _validate_operator(AssertionOperator["$="], "Hello Robots", "Hello"),
        _validate_operator(AssertionOperator["$="], "Hel[4,5]?[1-9]+ Robots", "[1-9]+ Robots"),
        _validate_operator(AssertionOperator["$="], "Hel[4,5]?[1-9]+ Robots", ".*"),
    ]
    verify_all("ends with", results)


def _validate_operator(operator: AssertionOperator, actual, expected):
    try:
        return verify_assertion(actual, operator, expected)
    except Exception as error:
        return error


def _method_validator(method, arg=None):
    try:
        if arg is None:
            return method()
        return method(arg)
    except Exception as error:
        return error
