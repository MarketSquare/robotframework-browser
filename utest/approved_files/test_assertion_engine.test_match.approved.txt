match

0) Actual
1) 'Actual' (str) should match '/(\d)+/' (str)
2) Actual
3) 'Actual' (str) should match '/(\d)+/' (str)
4) Actual
5) Actual
multiline
6) 'Actual
multiline' (str) should match '/(\d)+/' (str)
